src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,519 @@
+        log.debug("HEADER")
+        log.debug("------")
+        log.debug("protocol ID     {pid!r}", pid=protocol_id)
+        log.debug("size            {hs}", hs=hdr_size)
+        log.debug("credit charge   {cc}", cc=self.credit_charge)
+        log.debug("status          {status}", status=hdr_status)
+        log.debug("command         {cmd!r} {cmdn:02x}",
+                  cmd=COMMANDS[self.hdr_command],
+                  cmdn=self.hdr_command)
+        log.debug("credit request  {cr}", cr=self.credit_request)
+        s = ""
+        if self.is_async:
+            s += "ASYNC "
+        if self.is_signed:
+            s += "SIGNED "
+        if self.is_related:
+            s += "RELATED "
+        log.debug("flags           0x{flags:x} {s}", flags=self.hdr_flags, s=s)
+        log.debug("next command    0x{nc:x}", nc=self.next_command)
+        log.debug("message ID      0x{mid:x}", mid=self.message_id)
+        log.debug("session ID      0x{sid:x}", sid=self.session_id)
+        if self.is_async:
+            log.debug("async ID        0x{aid:x}", aid=self.async_id)
+        else:
+            log.debug("tree ID         0x{tid:x}", tid=self.tree_id)
+        log.debug("signature       {sig}",
+                  sig=binascii.hexlify(self.signature))
+        try:
+            name, req_type, resp_type = COMMANDS[self.hdr_command]
+        except IndexError:
+            log.error("unknown command 0x{cmd:x}", cmd=self.hdr_command)
+            self.error_response(STATUS_NOT_IMPLEMENTED)
+        else:
+            func = 'smb_' + name
+            try:
+                if hasattr(self, func) and req_type:
+                    req = req_type(packet[64:])
+                    resp = resp_type()
+                    getattr(self, func)(req, resp)
+                else:
+                    log.error("command '{cmd}' not implemented",
+                              cmd=COMMANDS[self.hdr_command][0])
+                    self.error_response(STATUS_NOT_IMPLEMENTED)
+            except base.SMBError as e:
+                log.error(str(e))
+                self.error_response(e.ntstatus)
+            except BaseException:
+                log.failure("in {cmd}", cmd=COMMANDS[self.hdr_command][0])
+                self.error_response(STATUS_UNSUCCESSFUL)
+        if self.is_related and self.next_command > 0:
+            self.packetReceived(packet[self.next_command:])
+
+    def send_with_header(self, payload, command=None, status=STATUS_SUCCESS):
+        """
+        prepare and transmit a SMB header and payload
+        so a full packet but focus of function on header construction
+
+        @param command: command name or id, defaults to same as received packet
+        @type command: L{str} or L{int}
+
+        @param payload: the after-header data
+        @type payload: L{bytes}
+
+        @param status: packet status, an NTSTATUS code
+        @type status: L{int}
+        """
+        payload = bytes(payload)
```

Comment:
If the payload is required to be bytes why is this necessary?

Reply:
it could also be a `GeneralStruct` which had a `__bytes__` method but as this is going away I will remove it